tests/std/tests/P0952R2_new_generate_canonical/test.cpp

```diff
@@ -0,0 +1,206 @@
+    assert((generate_canonical<float, 32>) (mt2) < 1.0f);
+    assert((generate_canonical<float, 32>) (mt2) < 1.0f);
+    assert((generate_canonical<float, 32>) (mt2) < 1.0f);
+}
+
+void test_tr1_16() {
+    using E_std = linear_congruential_engine<uint32_t, 75, 74, 65537>;
+    using E_tr1 = linear_congruential<uint32_t, 75, 74, 65537>;
+
+    E_std std_eng;
+    E_tr1 tr1_eng;
+
+    for (int i = 0; i < 10; ++i) {
+        assert(std_eng() == tr1_eng());
+    }
+
+    std_eng.seed();
+    tr1_eng.seed();
+
+    uniform_real_distribution<float> dist_32;
+    for (int i = 0; i < 10; ++i) {
+        assert(dist_32(std_eng) == dist_32(tr1_eng));
+    }
+
+    uniform_real_distribution<double> dist_64;
+    for (int i = 0; i < 10; ++i) {
+        assert(dist_64(std_eng) == dist_64(tr1_eng));
+    }
+}
+
+void test_tr1_32() {
+    using E_std = minstd_rand;
+    using E_tr1 = linear_congruential<unsigned int, 48271, 0, 2147483647>;
+
+    E_std std_eng;
+    E_tr1 tr1_eng;
+
+    for (int i = 0; i < 10; ++i) {
+        assert(std_eng() == tr1_eng());
+    }
+
+    std_eng.seed();
+    tr1_eng.seed();
+
+    uniform_real_distribution<float> dist_32;
+    for (int i = 0; i < 10; ++i) {
+        assert(dist_32(std_eng) == dist_32(tr1_eng));
+    }
+
+    uniform_real_distribution<double> dist_64;
+    for (int i = 0; i < 10; ++i) {
+        assert(dist_64(std_eng) == dist_64(tr1_eng));
+    }
+}
+
+void test_tr1_64() {
+    using E_std = mt19937_64;
+    using E_tr1 = mersenne_twister<unsigned long long, 64, 312, 156, 31, 0xb5026f5aa96619e9ULL, 29, 17,
+        0x71d67fffeda60000ULL, 37, 0xfff7eee000000000ULL, 43>;
+
+    E_std std_eng;
+    E_tr1 tr1_eng(E_tr1::default_seed, 0x5555555555555555ULL, 6364136223846793005ULL);
+
+    std_eng.seed();
+    tr1_eng.seed(E_tr1::default_seed, 6364136223846793005ULL);
+
+    for (int i = 0; i < 10; ++i) {
+        assert(std_eng() == tr1_eng());
+    }
+
+    std_eng.seed();
+    tr1_eng.seed(E_tr1::default_seed, 6364136223846793005ULL);
+
+    uniform_real_distribution<float> dist_32;
+    for (int i = 0; i < 10; ++i) {
+        assert(dist_32(std_eng) == dist_32(tr1_eng));
+    }
+
+    uniform_real_distribution<double> dist_64;
+    for (int i = 0; i < 10; ++i) {
+        assert(dist_64(std_eng) == dist_64(tr1_eng));
+    }
+}
+
+template <class Real, size_t Bits, class Engine>
+Real generate_with_ibe() {
+    independent_bits_engine<Engine, Bits, uint64_t> ibe;
+    return generate_canonical<Real, 64>(ibe);
+}
+
+int main() {
+    {
+        // edge case, Bits == 0
+        using Engine = ranlux24;
+        Engine eng;
+        assert((generate_canonical<double, 0, Engine>) (eng) == 0.0);
+        assert(eng() == Engine{}());
+    }
+
+    {
+        // float, URBG range is a power of two
+        using Engine                = ranlux24;
+        constexpr uint32_t values[] = {0xE57B2C, 0xF91555, 0xD9F2DE};
+        Engine eng;
+        for (size_t i = 0; i < std::size(values); ++i) {

```

Comment:
This test is already `using namespace std;`, but actually we can just use range-for here. Occurs below.